src/animated/Controller.ts

```diff
@@ -4,61 +4,84 @@ import {
   is,
   toArray,
   withDefault,
+  hasKeys,
 } from '../shared/helpers'
 import AnimatedValue from './AnimatedValue'
 import AnimatedValueArray from './AnimatedValueArray'
 import { start, stop } from './FrameLoop'
 import { colorNames, interpolation as interp, now } from './Globals'
+import { SpringProps } from '../../types/renderprops'
 
-type FinishedCallback = (finished?: boolean) => void
+type Animation = any
+
+type ControllerProps<DS extends object> = DS &
+  SpringProps<DS> & {
+    attach?: (ctrl: Controller) => Animation
+  }
 
 type AnimationsFor<P> = { [Key in keyof P]: any }
 
 type ValuesFor<P> = { [Key in keyof P]: any }
 
-type InterpolationsFor<P> = {
-  [Key in keyof P]: P[Key] extends ArrayLike<any>
+type InterpolationsFor<DS> = {
+  [K in keyof DS]: DS[K] extends ArrayLike<any>
     ? AnimatedValueArray
     : AnimatedValue
 }
 
-let G = 0
-class Controller<P extends any = {}> {
-  id: number
+type EndCallback = (finished?: boolean) => void
 
+let nextId = 1
+class Controller<DS extends object = any> {
+  id = nextId++
   idle = true
-  hasChanged = false
   guid = 0
-  local = 0
```

Comment:
I dropped the `local` property in favor of passing the `guid` through the `runAsync` method, since that's the only place the `local` property was being used.

Reply:
local took the current guid so that async springs can bail out if their frame is out of date, i guess passing it through is the same thing